src/main/java/net/openhft/chronicle/queue/impl/single/TSQueueLock.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2014-2017 Higher Frequency Trading
+ *
+ * http://www.higherfrequencytrading.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.openhft.chronicle.queue.impl.single;
+
+import net.openhft.chronicle.core.values.LongValue;
+import net.openhft.chronicle.queue.impl.TableStore;
+import net.openhft.chronicle.queue.impl.table.SingleTableBuilder;
+import net.openhft.chronicle.threads.Pauser;
+import net.openhft.lang.Jvm;
+
+import java.io.File;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+/**
+ * Implements queue lock via TableStore mechanism.
+ */
+public class TSQueueLock implements QueueLock {
+
+    private static final String QUEUE_LOCK_FILE = "queue-lock" + SingleTableBuilder.SUFFIX;
+    private static final String LOCK_KEY = "chronicle.queue.lock";
+    private static final long LOCK_WAIT_TIMEOUT = Long.getLong("chronicle.queue.lock.timeoutMS", 30_000);
+    private static final long UNLOCKED = Long.MIN_VALUE;
+    private static final long PID = Jvm.getProcessId();
+    private final LongValue lock;
+    private final Pauser pauser;
+    private final ThreadLocal<Long> lockHolderTidTL = new ThreadLocal<>();
+
+    public TSQueueLock(File queueDirectoryPath, Supplier<Pauser> pauser) {
+        final File storeFilePath;
+        if ("".equals(queueDirectoryPath.getPath())) {
+            storeFilePath = new File(QUEUE_LOCK_FILE);
+        } else {
+            storeFilePath = new File(queueDirectoryPath, QUEUE_LOCK_FILE);
+            queueDirectoryPath.mkdirs();
+        }
+        TableStore tableStore = SingleTableBuilder.binary(storeFilePath).build();
```

Comment:
I think this is prone to concurrent threads creating multiple instances of the LOCK_KEY reference in the table.

I had to use file locking to get around this for the directory listing: https://github.com/OpenHFT/Chronicle-Queue/blob/master/src/main/java/net/openhft/chronicle/queue/impl/single/TableDirectoryListing.java#L51

Reply:
Hmm I didn't check where exactly you did it, I assumed it was fixed for table store as a whole - why don't we move this to main TableStore implementation to avoid duplication of code?

Reply:
The pattern so far is to use a lock around initial calls to acquireKey, then just use the LongValue references. If the file-lock is moved to acquireKey() in table store, then it will be invoked each time. 
I think the change to move it there would be ok, but it may be worth overloading the method and calling it acquireKeyWithExclusiveLock or something similar.Hm. The pattern so far is to use a lock around initial calls to acquireKey, then just use the LongValue references. If the file-lock is moved to acquireKey() in table store, then it will be invoked each time. 
I think the change to move it there would be ok, but it may be worth overloading the method and calling it acquireKeyWithExclusiveLock or something similar.

Reply:
Changed to use locking